src/jit/compiler.cpp

```diff
@@ -516,64 +487,450 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
+            useType = TYP_FLOAT;
+        }
+        else
+        {
+            useType = TYP_INT;
+        }
         break;
-#endif // _TARGET_64BIT_
 
-    // Pointer size
-#ifdef _TARGET_64BIT_
-#ifndef _TARGET_AMD64_
+#ifndef _TARGET_XARCH_   
     case 5:
     case 6:
     case 7:
-#endif // _TARGET_AMD64_
-    case 8:
-#else // !_TARGET_64BIT_
-    case 4:
-#endif // !_TARGET_64BIT_
-        info.compCompHnd->getClassGClayout(clsHnd, &gcPtr);
-        useType = getJitGCType(gcPtr);
+        useType = TYP_I_IMPL;
+        break;
+
+#endif // _TARGET_XARCH_
+#endif // _TARGET_64BIT_
+
+    case TARGET_POINTER_SIZE:
+        if (IsHfa(clsHnd))
+        {
+#ifdef _TARGET_64BIT_
+            var_types hfaType = GetHfaType(clsHnd);
+            // Check and exclude the case of an HFA of two floats
+            if (hfaType == TYP_DOUBLE)
+            {
+                // We have an HFA of one double
+                useType = TYP_DOUBLE;
+            }  
+            else 
+            {
+                assert(hfaType == TYP_FLOAT);
+
+                // We have an HFA of two floats
+                // This should  be passed or returned in two FP registers
+                useType = TYP_UNKNOWN;
+            }
+#else  // 32BIT
+            // structSize of 4 with IsHfa, we must have an HFA of one float
+            useType = TYP_FLOAT;
+#endif
+        }
+        else
+        {
+            BYTE gcPtr = 0;
+            // Check if this pointer-sized struct is wrapping a GC object
+            info.compCompHnd->getClassGClayout(clsHnd, &gcPtr);
+            useType = getJitGCType(gcPtr);
+        }
         break;
 
     default:
-#if FEATURE_MULTIREG_RET
-        if (forReturn)
+        useType = TYP_UNKNOWN;
+        break;
+    }
+
+    return useType;
+}
+
+//-----------------------------------------------------------------------------
+// getArgTypeForStruct: 
+//     Get the type that is used to pass values of the given struct type.
+//     If you already have retrieved the struct size then it should be passed
+//     as the third optional argument as it allows us to avoid an extra call
+//     to getClassSize(clsHnd)
+//
+// Arguments:
+//    clsHnd       - the handle for the struct type
+//    wbPassStruct - information about how the struct is passed
+//    structSize   - the size of the struct type, 
+//                   or 0 if we should to call getClassSize(clsHnd)
+//
+// Return Value:
+//    When *wbPassStruct is SPK_PrimitiveType this method's return value:
+//       is the primitive type used to pass the struct.
+//    When *wbPassStruct is SPK_ByReference this method's return value:
+//       is TYP_UNKNOWN and the struct type is passed by reference to a copy
+//    When *wbPassStruct is SPK_ByValue or SPK_ByValueAsHfa the return value:
+//       can be TYP_STRUCT and the struct type is passed using multiple registers.
+//       can be TYP_UNKNOWN and the struct type is passed by value (x86 and ARM32 only)
```

Comment:
Right. I should have realized that this method doesn't have all the context to determine whether these will actually fit in registers (since it doesn't know where this is in the arg order). I see that now. But does there need to be a distinction between structs that COULD be passed by value in registers, if there are sufficient, and those that are too big to be passed in registers in any case?